[Lens] Suggestion improvements

flash1293 opened the pull request:
This PR improves suggestions to be more useful in standard scenarios:

* Introduce a `changeType` on data table suggestions describing what changed inn comparison to the currently active datasource state - can be used by visualizations to improve suggestions
* Introduce a `label` on data table suggestions which describes the table (can be used by the visualization to create the title)
* Make it possible to hide suggestions (not showing up in the suggestion panel, just used for directly applying from chart switcher or from field drop)
* Show 5 instead of 3 suggestions
* Apply filter/timerange context to the suggestions as well
* Remove isMetric on metadata but leave a TODO
* Make Average the default number field operation
* Make Count the default document operation

Suggestions got extended as follows:
* XY, table and metric visualizations use the datatable label if available
* XY visualizations suggest a slight presentational change (area <-> bar, flip horizontal) if the data table stays unchanged
* Indexpattern also suggests metric only tables
* Indexpattern suggests over default time field via date histogram if there is currently no time dimension
* Indexpattern suggests other metric as metric only table if current table is a bucketed table.

Things missing:

* Scale the metric further down in the preview to always show the hole number - I'm going to do this in a separate PR

flash1293 commented:
Dear reviewers, this PR is not completely finnished yet but it can already be review as the additional changes will be small.

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/13594/)

AlonaNadler commented:
This is so great @flash1293 !!!
huge progress on suggestions 👍 👍 


- @chrisdavies , I love the metric chart ❤️ 

few comments: 
- The suggestions move too much in the background, it is difficult to understand why they move and how to go back (look at the suggestion panel in the gif). 
![Aug-22-2019 14-29-23](https://user-images.githubusercontent.com/32550316/63551206-5e4c6600-c4e9-11e9-87bb-681863195e8c.gif)
I suggest the following: 
- suggestions should be generated when the user adds/removes a field explicitly or changes the metric function. In all other cases, the suggestion panel shouldn't change, refresh or load, only the preview panel.  The suggestion panel should change when switching between suggestions.
- the suggestion that shows in the preview should be highlighted in some form (cc @cchaos )that will infer that this is what presented in the preview and will help with orientation.
- the original preview should always be on the left side of the suggestions so users will know how to get back to it. Example: I dropped a field played with the suggestions and moved back to the original preview. 
-The tooltip on suggestions works well! 
- `Introduce a label on data table suggestions which describes the table (can be used by the visualization to create the title)` what's that? couldn't find it
- thanks for changing the default numerical function :)
- not sure related to suggestion but multiple y-series gets me to an empty chart
![image](https://user-images.githubusercontent.com/32550316/63555842-5abfdb80-c4f7-11e9-8ca8-16b94a67105e.png)
- on multiple y- series when using the suggestions I can't go back to my original 3 y-series chart
- it's gonna be so much fun to create charts like [this](https://demo.elastic.co/app/kibana#/visualize/edit/bfa5e400-1b16-11e7-b09e-037021c4f8df-ecs?_g=()&_a=(filters:!(),linked:!f,query:(language:kuery,query:''),uiState:(),vis:(aggs:!(),params:(axis_formatter:number,axis_position:left,axis_scale:normal,id:'32f46f40-1b16-11e7-b09e-037021c4f8df',index_pattern:'metricbeat-*',interval:auto,series:!((axis_position:right,chart_type:line,color:'rgba(211,49,21,1)',fill:'1',formatter:bytes,id:'4ff61fd0-1b16-11e7-b09e-037021c4f8df',label:Used,line_width:1,metrics:!((field:system.memory.actual.used.bytes,id:'4ff61fd1-1b16-11e7-b09e-037021c4f8df',type:avg)),point_size:'0',seperate_axis:0,split_mode:everything,stacked:stacked),(axis_position:right,chart_type:line,color:'rgba(0,156,224,1)',fill:'1',formatter:bytes,id:'753a6080-1b16-11e7-b09e-037021c4f8df',label:Cache,line_width:1,metrics:!((field:system.memory.actual.used.bytes,id:'753a6081-1b16-11e7-b09e-037021c4f8df',type:avg),(field:system.memory.used.bytes,id:'7c9d3f00-1b16-11e7-b09e-037021c4f8df',type:avg),(id:'869cc160-1b16-11e7-b09e-037021c4f8df',script:'params.actual+!!%3D+null+%26%26+params.used+!!%3D+null+%3F+params.used+-+params.actual+:+null',type:calculation,variables:!((field:'753a6081-1b16-11e7-b09e-037021c4f8df',id:'890f9620-1b16-11e7-b09e-037021c4f8df',name:actual),(field:'7c9d3f00-1b16-11e7-b09e-037021c4f8df',id:'8f3ab7f0-1b16-11e7-b09e-037021c4f8df',name:used)))),point_size:'0',seperate_axis:0,split_mode:everything,stacked:stacked),(axis_position:right,chart_type:line,color:%2368BC00,fill:'1',formatter:bytes,id:'32f46f41-1b16-11e7-b09e-037021c4f8df',label:Free,line_width:1,metrics:!((field:system.memory.free,id:'32f46f42-1b16-11e7-b09e-037021c4f8df',type:avg)),point_size:'0',seperate_axis:0,split_mode:everything,stacked:stacked)),show_grid:1,show_legend:1,time_field:'@timestamp',type:timeseries),title:'Memory+Usage+%5BMetricbeat+System%5D+ECS',type:metrics))) with lens 
- sometimes the preview is empty but suggestions has previews which is weird not sure why since there is data
![image](https://user-images.githubusercontent.com/32550316/63556157-9dce7e80-c4f8-11e9-82f0-8f38145ef10e.png)
- I select a field it does timeseries and uses avg, then in the suggestions it moves to be sum, I think the first single metric suggestion should be based on the preview metric, if we do another single metric it will be with sum. 
- when configurator has a split series have another suggestion that switched between the nesting of the x-axis and split series field

flash1293 commented:
Hi @AlonaNadler  , thanks for your feedback

> suggestions should be generated when the user adds/removes a field explicitly or changes the metric function. In all other cases, the suggestion panel shouldn't change, refresh or load, only the preview panel. The suggestion panel should change when switching between suggestions.

This is already how it's implemented. The problem here is that the suggestions change with a two second delay from the actual change. We did this to improve performance and avoid a lot of unnecessary data fetching. There is definitely room for improvement here, and it's already on the list. Besides that we could also show the new suggestions instantly but only start fetching the expression after a second or so and show a loading spinner till then. By doing so we eliminate the case of suggestions changing when the users wants to click them. What do you think @chrisdavies @timroes ? Anyway to avoid blowing up this huge PR even further, I will move this to a second one.

> the suggestion that shows in the preview should be highlighted in some form (cc @cchaos )that will infer that this is what presented in the preview and will help with orientation.

The exact vis from the preview should never show in the suggestion panel. Do you mean suggestions based on the same visualization?

> the original preview should always be on the left side of the suggestions so users will know how to get back to it. Example: I dropped a field played with the suggestions and moved back to the original preview.

That's not possible currently because applying a suggestion may result in loss of configuration - e.g. if you switch from an xy chart with two dimension to a metric chart with only the metric dimension, the x dimension is lost and it's not possible to switch back. We already talked about this in context of the chart switcher - maybe something for after beta?

> Introduce a label on data table suggestions which describes the table (can be used by the visualization to create the title) what's that? couldn't find it

This is just a not about the internal data flow not visible to the user.

> not sure related to suggestion but multiple y-series gets me to an empty chart

> sometimes the preview is empty but suggestions has previews which is weird not sure why since there is data

Maybe that's because it's an area chart now and you only have a single data point because your time range is 15 mins and the interval defaults to `1d` (I'm guessing the 15mins part)? This should work better with `auto` date histogram.

> I select a field it does timeseries and uses avg, then in the suggestions it moves to be sum, I think the first single metric suggestion should be based on the preview metric, if we do another single metric it will be with sum.

That was done on purpose, I misinterpreted the table here. I will change that 👍 

> when configurator has a split series have another suggestion that switched between the nesting of the x-axis and split series field

Good idea, I will add that.

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/13976/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/14020/)

flash1293 commented:
Jenkins, test this.

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/14029/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/14050/)

AlonaNadler commented:
> This is already how it's implemented.

The suggestions seem to change when I click on one of them, I doesn look like only loading but different chart type etcc. See in the gif, when I select one of the suggestions the other suggestions are changing 
![Aug-23-2019 16-33-03](https://user-images.githubusercontent.com/32550316/63628872-c53e4d80-c5c3-11e9-86a0-4a3bd98ea16b.gif)


> Besides that we could also show the new suggestions instantly but only start fetching the expression after a second or so and show a loading spinner till then.

Great idea, loading while the suggestions are running will be good 

> The exact vis from the preview should never show in the suggestion panel. Do you mean suggestions based on the same visualization?

Currently, if I selected one of the suggestions, I'm not sure how to go back to the original chart I had in the preview. 
To solve that I think we can have the original preview as the first on the left and show it is selected to infer this is what the user sees in the preview, then if I select another suggestion the new suggested will be highlighted and also I can easily go back to the preview as it is always on further to the left. 

> That's not possible currently because applying a suggestion may result in loss of configuration - e.g. if you switch from a xy chart with two dimensions to a metric chart with only the metric dimension, the x dimension is lost and it's not possible to switch back. We already talked about this in context of the chart switcher - maybe something for after beta?

I remember that if you had the original preview as the first suggestion wouldn't that be less of a problem? since I will always have a way to go back to where I started


> Good idea, I will add that.

Thanks :)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/14262/)

flash1293 commented:
Adjusted as follows:
* Keep current metric operation when suggesting bucket-less variant of current chart
* Suggest other metric when currently single metric is shown
* Suggest changing the nesting for two-bucket charts

flash1293 commented:
@AlonaNadler Did we also talk about having a suggestion for stacked/non-stacked if applicable? Not sure anymore

flash1293 commented on x-pack/legacy/plugins/lens/public/datatable_visualization_plugin/visualization.tsx:
```diff
@@ -127,42 +127,57 @@ export const datatableVisualization: Visualization<
-    });
+            previewIcon: 'visTable',
+            // dont show suggestions for reduced versions or single-line tables
+            hide: table.changeType === 'reduced' || !table.isMultiRow,
```
If we do this, the single-row table won't be re-used if you switch e.g. from a metric visualization to a table visualization. There is a point to drop the metric if you do that, but I think it's nice to carry over as much as possible.

flash1293 commented on x-pack/legacy/plugins/lens/public/editor_frame_plugin/editor_frame/suggestion_panel.tsx:
```diff
@@ -162,3 +163,35 @@ function InnerSuggestionPanel({
+  datasourceStates: Record<string, { isLoading: boolean; state: unknown }>,
+  framePublicAPI: FramePublicAPI
+) {
+  if (!suggestion.previewExpression) return null;
```
Moved, but it didn't really make things much simpler... The problem is that there are a lot of conditions that can cause the expression to become null, but getting rid of one of them is probably worth it.

flash1293 commented on x-pack/legacy/plugins/lens/public/indexpattern_plugin/indexpattern_suggestions.ts:
```diff
@@ -49,10 +50,15 @@ function buildSuggestion({
+        columnId,
+        operation: columnToOperation(columnMap[columnId]),
+      })),
+      isMultiRow: typeof isMultiRow === 'undefined' || isMultiRow,
```
That's not what this check does - if `isMultiRow` is not defined, this will be true, otherwise it will be the value of `isMultiRow`. But I think this is highlighting that the statement is hard to parse. I removed `isMultiRow` and just look at the columns in `buildSuggestion`

flash1293 commented on x-pack/legacy/plugins/lens/public/indexpattern_plugin/indexpattern_suggestions.ts:
```diff
@@ -358,10 +495,49 @@ function createSimplifiedTableSuggestions(state: IndexPatternPrivateState, layer
+      return buildSuggestion({
+        state,
+        layerId,
+        isMultiRow: updatedLayer.columnOrder.length > 1,
```
In general yes, but we only get into this branch with single column metric tables and multi column bucket + metric tables. However I agree that this is confusing and moved the check to `buildSuggestion`

flash1293 commented on x-pack/legacy/plugins/lens/public/xy_visualization_plugin/xy_suggestions.ts:
```diff
@@ -92,11 +113,98 @@ function prioritizeColumns(columns: TableSuggestionColumn[]) {
+  };
+
+  // if current state is using the same data, suggest same chart with different presentational configuration
+  if (currentState && changeType === 'unchanged') {
```
Thanks for these, I agree they make the functions easier to parse 👍

flash1293 commented on x-pack/legacy/plugins/lens/public/xy_visualization_plugin/xy_config_panel.tsx:
```diff
@@ -28,7 +28,7 @@ import { NativeRenderer } from '../native_renderer';
 import { generateId } from '../id_generator';
 
-const isNumericMetric = (op: OperationMetadata) => op.isMetric && op.dataType === 'number';
+const isNumericMetric = (op: OperationMetadata) => !op.isBucketed && op.dataType === 'number';
```
Agree, I just wasn't sure where to put it yet - didn't want to start another top level directory or a `utils` folder.

flash1293 commented on x-pack/legacy/plugins/lens/public/editor_frame_plugin/editor_frame/chart_switch.tsx:
```diff
@@ -112,6 +112,10 @@ export function ChartSwitch(props: Props) {
+      // don't use extended versions of current data table on switching between visualizations
+      // to avoid confusing the user.
+      return suggestion.changeType !== 'extended';
     })[0];
```
That could also happen in the old version of the code (e.g. if the datasource doesn't return any suggestions) and is already handled - the return object checks whether `topSuggestion` exists before acting on it.

To make this more explicit I added a cast to reflect it in the types:
```ts
/// ...
    })[0] as Suggestion | undefined;
```

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/396/)
